### ISSUE TITLE
meson assertion when restoreMemoryProfile=true

### ISSUE THREAD

**Issue by mmphys** [aportelli/Hadrons#23]
I can run a job from a database using xml such as the attached:
[Z2_s.3080.xml.txt](https://github.com/aportelli/Hadrons/files/4804664/Z2_s.3080.xml.txt)
However, when I set restoreMemoryProfile=true, the job fails in the first meson contraction with an error such as:
Hadrons : Message  : 7747.190612 s : ---------------- Measurement step 461/11249 (module 'meson_3pt_s_GFPW_quark_h0_h2_gT5_dt_20_p_0_0_0_t_16') ----------------
Hadrons : Message  : 7747.190628 s : ................ Module execution
Hadrons : Message  : 7747.190667 s : Computing meson contractions 'meson_3pt_s_GFPW_quark_h0_h2_gT5_dt_20_p_0_0_0_t_16' using quarks 'Prop_GFPW_h2_p_0_0_0_t_16' and 'PropSeq_GFPW_h0_gT5_dt_20_ps_0_0_0_s_p_0_0_0_t_16'
HadronsXmlRun: /home/dp008/dp008/dc-mars3/.local/Double/include/Grid/lattice/Lattice_base.h:50: void Grid::conformable(Grid::GridBase *, Grid::GridBase *): Assertion `lhs == rhs' failed.
The backtrace shows this happens at Meson.hpp:244, i.e.
    buf = sink(c);
This happens repeatably for jobs with circa 10k modules. I did notice this once during testing on my laptop, but assumed it was user error as it did not recur.

**Comment by aportelli** [comment]
Could I have the DB please?

**Comment by mmphys** [comment]
Here are two very similar dBs, but with different schedule, both exhibit the issue.
[db.tar.gz](https://github.com/aportelli/Hadrons/files/4804795/db.tar.gz)

**Comment by mmphys** [comment]
Grid Commit: 0174f5f742782d1b43e49213bd9d729f7094962e [0174f5f7]
Hadrons Commit: 04e06e8f0dc5f1460197a118c966f5a1047be834 [04e06e8]
i.e. latest Grid, but Hadrons is prior to Fionn's latest changes and HadronsXmlValidate fix

**Comment by aportelli** [comment]
I can reproduce it. The error you get is about Grid pointers being inconsistent, which the DB knows nothing about (it does not save the geometries)... this is weird let me investigate.

**Comment by aportelli** [comment]
Ufff that was a painful one... but it was a bug related to the DB indeed, so good catch!
The error message really did not suggest anything like that and was a randomly appearing, indirect side-effect of the problem.

The short version is: the DB was not restoring the storage type (standard, cache or tmp) of the objects. The momentum phase in the sink function got 'standard' by default instead of 'cache', which means it will be part of garbage collection. Because it is not needed by anybody it was destroyed as soon as the sink module ended. Later in the meson contraction the sink function is called, and randomly it was possible to dereference the pointer on the destroyed object without triggering a bad access error.

So one additional issue is that the sink function was just capturing the address of the phase, which is fine but unsafe if the phase get destroyed in the mean time. So once more it is very important to use the `envGet` macro as it actually checks the object lives. This is a general comment I actually wrote the unsafe code 😄.

It is now fixed in `develop`, the DB restore the storage types and the phases are accessed in a safer way (i.e. the bug would result in a meaningful error message). Let me know if it works on your side.

This would have happened to us in production at one point or another, so thanks for the thorough testing it saved us a lot of headache.
